server/src/main/java/org/elasticsearch/common/filesystem/FileSystemNatives.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.filesystem;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.util.Constants;
+
+import java.nio.file.Path;
+import java.util.OptionalLong;
+
+/**
+ * This class provides utility methods for calling some native methods related to filesystems.
+ */
+public final class FileSystemNatives {
+
+    private static final Logger logger = LogManager.getLogger(FileSystemNatives.class);
+
+    interface Provider {
+        OptionalLong allocatedSizeInBytes(Path path);
+    }
+
+    private static final class NoopFileSystemNativesProvider implements Provider {
+        @Override
+        public OptionalLong allocatedSizeInBytes(Path path) {
+            return OptionalLong.empty();
+        }
+    }
+
+    private static final Provider JNA_PROVIDER;
+    static {
```

Comment:
Consider extracting the loading of the provider to a method, so the control flow becomes more clear and we don't need use `finally` to assign the value to the final variable.

```java
 private static final Provider JNA_PROVIDER = loadJnaProvider();
private static Provider loadJnaProvider() {
        try {
            // load one of the main JNA classes to see if the classes are available. this does not ensure that all native
            // libraries are available, only the ones necessary by JNA to function
            Class.forName("com.sun.jna.Native");
            if (Constants.WINDOWS) {
                return WindowsFileSystemNatives.getInstance();
            }
        } catch (ClassNotFoundException e) {
            logger.warn("JNA not found. FileSystemNatives methods will be disabled.", e);
        } catch (LinkageError e) {
            logger.warn("unable to load JNA native support library, FileSystemNatives methods will be disabled.", e);
        }
        return new NoopFileSystemNativesProvider();
    }
```

Reply:
Sure, I pushed 9258da1a4fe7cbaa6f5bbbbfd1ee303a383bf4b4